File: frontend/src/api/interviews.ts
import { apiClient } from "./client";

export interface InterviewAnswer {
  id: string;
  question_id: string;
  answer_text: string;
  ai_feedback?: string;
  study_recommendations?: { topic: string; description: string }[];
  reference_links?: { url: string; title: string; excerpt: string }[];
  attempt_number: number;
  created_at: string;
}

export interface InterviewQuestion {
  id: string;
  question_number: number;
  question_type: "resume" | "values" | "technical" | "situational";
  question_text: string;
  hint_text?: string;
  reference_links?: { url: string; title: string; excerpt: string }[];
  is_follow_up: boolean;
  parent_question_id?: string;
  points_consumed: number;
  follow_up_count: number;
  answers: InterviewAnswer[];
  created_at: string;
}

export interface InterviewSession {
  id: string;
  project_id: string;
  user_id: string;
  status: "generating" | "ready" | "in_progress" | "done";
  total_questions: number;
  total_follow_ups: number;
  total_points_spent: number;
  created_at: string;
  completed_at?: string;
  questions: InterviewQuestion[];
  company_name?: string;
  position?: string;
}

export interface InterviewSessionListItem {
  id: string;
  project_id: string;
  status: string;
  total_questions: number;
  total_points_spent: number;
  created_at: string;
  completed_at?: string;
  company_name?: string;
  position?: string;
}

export interface SessionSummary {
  overall_score: number;
  strengths: string[];
  weaknesses: string[];
  improvement_tips: string[];
  study_plan: { topic: string; description: string }[];
  question_scores: { question: string; score: number; summary: string }[];
}

export const QUESTION_TYPE_LABELS: Record<string, string> = {
  resume: "자소서 기반",
  values: "가치관",
  technical: "기술/직무",
  situational: "상황 대처",
};

export const interviewApi = {
  startSession: (projectId: string) =>
    apiClient
      .post<InterviewSession>(`/projects/${projectId}/interviews`, {})
      .then((r) => r.data),

  listSessions: (projectId: string) =>
    apiClient
      .get<InterviewSessionListItem[]>(`/projects/${projectId}/interviews`)
      .then((r) => r.data),

  getSession: (projectId: string, sessionId: string) =>
    apiClient
      .get<InterviewSession>(`/projects/${projectId}/interviews/${sessionId}`)
      .then((r) => r.data),

  submitAnswer: (projectId: string, sessionId: string, questionId: string, answerText: string) =>
    apiClient
      .post<InterviewAnswer>(`/projects/${projectId}/interviews/${sessionId}/answer`, {
        question_id: questionId,
        answer_text: answerText,
      })
      .then((r) => r.data),

  requestFollowUp: (projectId: string, sessionId: string, parentQuestionId: string) =>
    apiClient
      .post<InterviewQuestion>(`/projects/${projectId}/interviews/${sessionId}/follow-up`, {
        parent_question_id: parentQuestionId,
      })
      .then((r) => r.data),

  requestNewQuestion: (projectId: string, sessionId: string) =>
    apiClient
      .post<InterviewQuestion>(`/projects/${projectId}/interviews/${sessionId}/new-question`)
      .then((r) => r.data),

  completeSession: (projectId: string, sessionId: string) =>
    apiClient
      .post<SessionSummary>(`/projects/${projectId}/interviews/${sessionId}/complete`)
      .then((r) => r.data),
};

File: backend/app/api/v1/interviews.py
from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.dependencies import get_current_user_id
from app.schemas.interview import (
    SessionResponse, SessionListItem, SubmitAnswerRequest,
    FollowUpRequest, QuestionResponse, AnswerResponse,
    SessionSummaryResponse,
)
from app.services.interview_service import InterviewService

router = APIRouter(prefix="/projects/{project_id}/interviews", tags=["interviews"])


@router.post("", response_model=SessionResponse, status_code=201)
async def start_session(
    project_id: str,
    user_id: str = Depends(get_current_user_id),
    db: AsyncSession = Depends(get_db),
):
    """면접 세션 시작 (60P 차감 + 초기 5개 질문 AI 생성)."""
    return await InterviewService(db).start_session(user_id, project_id)


@router.get("", response_model=list[SessionListItem])
async def list_sessions(
    project_id: str,
    user_id: str = Depends(get_current_user_id),
    db: AsyncSession = Depends(get_db),
):
    """프로젝트의 면접 세션 목록."""
    return await InterviewService(db).list_sessions(user_id, project_id)


@router.get("/{session_id}", response_model=SessionResponse)
async def get_session(
    project_id: str,
    session_id: str,
    user_id: str = Depends(get_current_user_id),
    db: AsyncSession = Depends(get_db),
):
    """면접 세션 상세 (질문 + 답변 포함)."""
    return await InterviewService(db).get_session(user_id, project_id, session_id)


@router.post("/{session_id}/answer", response_model=AnswerResponse)
async def submit_answer(
    project_id: str,
    session_id: str,
    body: SubmitAnswerRequest,
    user_id: str = Depends(get_current_user_id),
    db: AsyncSession = Depends(get_db),
):
    """답변 제출 → AI 피드백 반환."""
    return await InterviewService(db).submit_answer(
        user_id, project_id, session_id, str(body.question_id), body.answer_text,
    )


@router.post("/{session_id}/follow-up", response_model=QuestionResponse)
async def request_follow_up(
    project_id: str,
    session_id: str,
    body: FollowUpRequest,
    user_id: str = Depends(get_current_user_id),
    db: AsyncSession = Depends(get_db),
):
    """꼬리 질문 요청 (1P, 질문당 최대 5개)."""
    return await InterviewService(db).request_follow_up(
        user_id, project_id, session_id, str(body.parent_question_id),
    )


@router.post("/{session_id}/new-question", response_model=QuestionResponse)
async def request_new_question(
    project_id: str,
    session_id: str,
    user_id: str = Depends(get_current_user_id),
    db: AsyncSession = Depends(get_db),
):
    """신규 질문 요청 (3P)."""
    return await InterviewService(db).request_new_question(user_id, project_id, session_id)


@router.post("/{session_id}/complete", response_model=SessionSummaryResponse)
async def complete_session(
    project_id: str,
    session_id: str,
    user_id: str = Depends(get_current_user_id),
    db: AsyncSession = Depends(get_db),
):
    """세션 종료 → AI 종합 분석."""
    return await InterviewService(db).complete_session(user_id, project_id, session_id)
